### PR TITLE
Create a fork manifest to store any fork schedules.

### PR DESCRIPTION
**Pull request by rolfyone**
Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    final UInt64 genesisEpoch = UInt64.valueOf(genesisConstants.getGenesisEpoch());
+    forkSchedule.add(new Fork(genesisForkVersion, genesisForkVersion, genesisEpoch));
+  }
+
+  @VisibleForTesting
+  void addFork(final Fork fork) {
+    final Fork lastFork = forkSchedule.get(forkSchedule.size() - 1);
+    checkArgument(fork.getEpoch().isGreaterThan(lastFork.getEpoch()));
+    forkSchedule.add(fork);
+  }
```
Comment (mbaxter) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java]: Would be nice to make `ForkManifest` immutable.  A few options for doing that:
* Make the constructor private, accepting a list of `Fork`'s, add some static factory methods:
  * `create(SpecConstants genesisConstants)`
  * `create(List<Fork> forks)`
* Create a `Fork.Builder` class

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+import tech.pegasys.teku.datastructures.state.Fork;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.constants.SpecConstants;
+import tech.pegasys.teku.ssz.SSZTypes.Bytes4;
+
+public class ForkManifest {
+  private final List<Fork> forkSchedule;
+
+  public ForkManifest(SpecConstants genesisConstants) {
```
Comment (mbaxter) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java]: Assuming we make this constructor private, accepting a list of forks (suggestion below), we could add some more validation here to make sure the forks are consistent:
* Ensure that each `fork.previousVersion == prevFork.currentVersion`
* Ensure all `fork.epoch` values are unique
* Ensure we have a genesis fork
* Ensure `fork.currentVersion != fork.previousVersion` except for the genesis fork

Also, I don't suppose the list of forks is large, but we could keep them in sorted order in a `NavigableMap` (epoch -> fork).  Then we can query for an epoch's fork with `NavigableMap.floorEntry`.

**Review by mbaxter — COMMENTED** [review]
Looks good!  Just a few suggestions on `ForkManifest`.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java [hunk 3]
```diff
@@ -15,44 +15,55 @@
...
-    checkArgument(fork.getEpoch().isGreaterThan(lastFork.getEpoch()));
-    forkSchedule.add(fork);
+  public static ForkManifest create(final List<Fork> forkList) {
+    final NavigableMap<UInt64, Fork> schedule = new TreeMap<>();
+    forkList.stream()
+        .forEach(
+            fork -> {
+              if (!schedule.isEmpty()) {
+                final Fork lastFork = schedule.lastEntry().getValue();
+                checkArgument(lastFork.getEpoch().isLessThan(fork.getEpoch()));
```
Comment (mbaxter) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java]: (nit) Might be worth adding error strings here.  We can make the tests more specific by checking for these error strings as well.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java [hunk 4]
```diff
@@ -15,44 +15,55 @@
...
   }
 
-  @VisibleForTesting
-  void addFork(final Fork fork) {
-    final Fork lastFork = forkSchedule.get(forkSchedule.size() - 1);
-    checkArgument(fork.getEpoch().isGreaterThan(lastFork.getEpoch()));
-    forkSchedule.add(fork);
+  public static ForkManifest create(final List<Fork> forkList) {
+    final NavigableMap<UInt64, Fork> schedule = new TreeMap<>();
+    forkList.stream()
```
Comment (mbaxter) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java]: (nit) Probably more future-proof to centralize validation in the constructor

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java [hunk 5]
```diff
@@ -15,44 +15,55 @@
...
+        .forEach(
+            fork -> {
+              if (!schedule.isEmpty()) {
+                final Fork lastFork = schedule.lastEntry().getValue();
+                checkArgument(lastFork.getEpoch().isLessThan(fork.getEpoch()));
+                checkArgument(lastFork.getCurrent_version().equals(fork.getPrevious_version()));
+                checkArgument(!fork.getPrevious_version().equals(fork.getCurrent_version()));
+              }
```
Comment (mbaxter) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java]: I think we need add an `else` clause here to assert that the first fork has an epoch == 0.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.TreeMap;
+import tech.pegasys.teku.datastructures.state.Fork;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.constants.SpecConstants;
+import tech.pegasys.teku.ssz.SSZTypes.Bytes4;
+
+public class ForkManifest {
+  private final NavigableMap<UInt64, Fork> forkSchedule;
+
+  private ForkManifest(final NavigableMap<UInt64, Fork> forkSchedule) {
```
Comment (mbaxter) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/ForkManifest.java]: We should also probably assert that `!forkSchedule.isEmpty()`

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/ForkManifestTest.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  void shouldGetMiddleFork() {
+    assertThat(forkManifest.get(UInt64.valueOf(100L))).isEqualTo(secondFork);
+    assertThat(forkManifest.get(UInt64.valueOf(199L))).isEqualTo(secondFork);
+  }
+
+  @Test
+  void shouldGetLastForkFromSchedule() {
+    assertThat(forkManifest.get(UInt64.valueOf(200L))).isEqualTo(thirdFork);
+    assertThat(forkManifest.get(UInt64.MAX_VALUE)).isEqualTo(thirdFork);
+  }
```
Comment (mbaxter) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/ForkManifestTest.java:109]: We should add a couple extra tests here:
* Try to create a `ForkManifest` by passing in an empty list
* Try to create a `ForkManifest` with only the 2nd and 3rd fork (no genesis fork)

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/ForkManifestTest.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  public void setup() {
+    when(constants.getGenesisForkVersion()).thenReturn(genesisForkVersion);
+    when(constants.getGenesisEpoch()).thenReturn(0L);
+    forkManifest = ForkManifest.create(List.of(genesisFork, secondFork, thirdFork));
+  }
+
+  @Test
+  void shouldNotAllowAppendingEarlierFork() {
+    assertThrows(
+        IllegalArgumentException.class, () -> ForkManifest.create(List.of(thirdFork, secondFork)));
```
Comment (mbaxter) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/ForkManifestTest.java]:
```suggestion
        IllegalArgumentException.class, () -> ForkManifest.create(List.of(genesisFork, thirdFork, secondFork)));
```

File: ethereum/spec/src/test/java/tech/pegasys/teku/spec/ForkManifestTest.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+  @BeforeEach
+  public void setup() {
+    when(constants.getGenesisForkVersion()).thenReturn(genesisForkVersion);
+    when(constants.getGenesisEpoch()).thenReturn(0L);
+    forkManifest = ForkManifest.create(List.of(genesisFork, secondFork, thirdFork));
+  }
+
+  @Test
+  void shouldNotAllowAppendingEarlierFork() {
```
Comment (mbaxter) [inline ethereum/spec/src/test/java/tech/pegasys/teku/spec/ForkManifestTest.java]:
```suggestion
  void shouldNotAllowUnorderedForks() {
```

**Review by ajsutton — APPROVED** [review]
LGTM.

**Pull request merged**